framework/src/task_manager/task_manager_core.c

```diff
@@ -986,81 +1004,83 @@ static int taskmgr_set_msg_cb(int type, void *data, int pid)
 	if (data == NULL) {
 		return TM_INVALID_PARAM;
 	}
+
 	handle = taskmgr_get_handle_by_pid(pid);
 	if (handle == TM_UNREGISTERED_APP) {
 		return handle;
 	}
-	if (type == TYPE_UNICAST) {
-		TM_UNICAST_CB(handle) = (tm_unicast_callback_t)data;
-	} else {
-		if (!taskmgr_is_tm_broadcast_msg_init()) {
-			taskmgr_broadcast_msg_init();
-		}
-		ret = taskmgr_check_broad_msg(((tm_broadcast_info_t *)data)->msg);
-		if (ret == TM_UNREGISTERED_MSG) {
-			return ret;
-		}
-		broadcast_info = taskmgr_search_broadcast_info(((tm_broadcast_info_t *)data)->msg, handle);
+
+	if (!taskmgr_is_tm_broadcast_msg_init()) {
+		taskmgr_broadcast_msg_init();
+	}
+
+	ret = taskmgr_check_broad_msg(data->msg);
+	if (ret == TM_UNREGISTERED_MSG) {
+		return ret;
+	}
+
+	broadcast_info = taskmgr_search_broadcast_info(data->msg, handle);
+	if (broadcast_info == NULL) {
+		broadcast_info = (tm_broadcast_info_t *)TM_ALLOC(sizeof(tm_broadcast_info_t));
 		if (broadcast_info == NULL) {
-			broadcast_info = (tm_broadcast_info_t *)TM_ALLOC(sizeof(tm_broadcast_info_t));
-			if (broadcast_info == NULL) {
+			return TM_OUT_OF_MEMORY;
+		}
+		broadcast_info->flink = NULL;
+		broadcast_info->msg = data->msg;
+		broadcast_info->cb = data->cb;
+		if (data->cb_data != NULL) {
+			broadcast_info->cb_data = TM_ALLOC(sizeof(tm_msg_t));
+			if (broadcast_info->cb_data == NULL) {
 				return TM_OUT_OF_MEMORY;
 			}
-			broadcast_info->flink = NULL;
-			broadcast_info->msg = ((tm_broadcast_info_t *)data)->msg;
-			broadcast_info->cb = ((tm_broadcast_info_t *)data)->cb;
-			if (((tm_broadcast_info_t *)data)->cb_data != NULL) {
-				broadcast_info->cb_data = TM_ALLOC(sizeof(tm_msg_t));
-				if (broadcast_info->cb_data == NULL) {
+			if (data->cb_data->msg_size != 0) {
+				broadcast_info->cb_data->msg = TM_ALLOC(data->cb_data->msg_size);
+				if (broadcast_info->cb_data->msg == NULL) {
 					return TM_OUT_OF_MEMORY;
```

Comment:
I think broadcast_info->cb_data should be freed when fail case

Reply:
Done.